doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,274 @@
+# Overview
+
+*Allow users to transform a source before persisting it using SQL*
+
+# Goals
+
+- Make append-only sources viable by retaining a bounded number of updates
+- Thin-down sources by filtering before consuming storage
+
+This design doc proposes supporting arbitrary data transformations during ingestion made available
+to users by extending the syntax of the `CREATE SOURCE` statement. The goal is to allow expressing
+any transformation that can be written as a SQL query with one candidate syntax being a `TRANSFORM
+USING <query>` fragment that can be optionally specified at the end of a `CREATE SOURCE` statement.
+
+**Example**
+
+```sql
+CREATE SOURCE foobar FROM POSTGRES ...
+TRANSFORM USING SELECT colA, colB FROM foobar WHERE colC > 4;
+```
+
+**Example with temporal filter**
+
+```sql
+CREATE SOURCE foobar FROM POSTGRES ...
+TRANSFORM USING SELECT colA, colB FROM foobar WHERE created_at > mz_now()
+```
+
+### Non-goals
+
+- Express existing envelopes and decoders in terms of a transformation query instead of custom code.
+  It seems possible but it’s out of scope for the initial implementation of this feature
+
+## Product Details
+
+While the proposed syntax allows arbitrary queries to be written there is a wide spectrum of product
+behavior that affects the kind of queries that the system is willing to accept and the promises to
+the user we are willing to make as to the effect of the queries will have to their cost model.
+
+Engineering wise, it is possible to support arbitrary queries (described below) where the system
+makes a best effort to push as much of that query as possible before we start persisting data. This
+is not always possible and in the worst case the provided query can result in zero gains in the
+amount of persisted data. For this reason we might want to constraint kind of queries we accept to
```

Comment:
I think it's even worse: can't users write a query that expands the amount of data that we store?